### ISSUE TITLE
StandaloneDeriving scoping and bracketing issues

### ISSUE THREAD

**Issue by Icelandjack** [RyanGlScott/ghc#7]
```haskell
import Data.Functor.Compose

newtype App f a = App (f a)
  deriving newtype
    (Functor, Applicative)

instance (Applicative f, Semigroup a) => Semigroup (App f a) where
  (<>) = liftA2 (<>)

deriving via (App (Compose f g) a) instance (Applicative f, Applicative g, Semigroup a) => Semigroup (Compose f g a)
```

doesn't scope over variables right (`f1` vs `f_apeW`) and there are some weird bracketing issues (`App Compose f g a` should be `App (Compose f g) a`)

```haskell
  instance (Applicative f1, Applicative g1,
            Semigroup a1) =>
           Semigroup (Compose f1 g1 a1) where
    (<>)
      = coerce
          @(App Compose f_apeW g_apeX a_apeY
            -> App Compose f_apeW g_apeX a_apeY
               -> App Compose f_apeW g_apeX a_apeY)
          @(Compose f_apeZ g_apf0 a_apf1
            -> Compose f_apeZ g_apf0 a_apf1
               -> Compose f_apeZ g_apf0 a_apf1)
          (<>)
```

**Comment by RyanGlScott** [comment]
I've known about this issue, but alas, it is not terribly straightforward to fix. Essentially, the renamer's treatment of `deriving via` is quite subtly broken. Due to a lucky fluke, it happens to work in the case of `deriving` clauses (except if you push it too hard, as in #4), but our luck runs out with `StandaloneDeriving`.

I believe that fixing this would likely require a fixing GHC Trac tickets [#14331](https://ghc.haskell.org/trac/ghc/ticket/14331) and [#14332](https://ghc.haskell.org/trac/ghc/ticket/14332) first, which is not exactly a quick endeavor. Let me know if this is blocking you and I can try to focus on that at some point, but given that most of our examples work with `deriving` clauses, I've put this on the backburner for now.

**Comment by RyanGlScott** [comment]
Actually, I lied—I think this is quite possible to fix independently of Trac [#14331](https://ghc.haskell.org/trac/ghc/ticket/14331) and [#14332](https://ghc.haskell.org/trac/ghc/ticket/14332). Patch incoming.

**Comment by RyanGlScott** [comment]
Also, it's worth noting that the parenthesization issue you encountered is a proper GHC bug that can be triggered without `deriving via`:

```haskell
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# OPTIONS_GHC -ddump-deriv #-}
module Bug where

import Control.Applicative
import Data.Functor.Compose
import Data.Semigroup

newtype App f a = MkApp (f a)
  deriving (Functor, Applicative)

instance (Applicative f, Semigroup a) => Semigroup (App f a) where
  (<>) = liftA2 (<>)

newtype Wat f g a = MkWat (App (Compose f g) a)
  deriving Semigroup
```
```
$ /opt/ghc/8.2.2/bin/ghci Bug.hs -dsuppress-uniques
GHCi, version 8.2.2: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

==================== Derived instances ====================
Derived class instances:

  <elided for brevity>

  instance (Data.Semigroup.Semigroup a, GHC.Base.Applicative g,
            GHC.Base.Applicative f) =>
           Data.Semigroup.Semigroup (Bug.Wat f g a) where
    (Data.Semigroup.<>)
      = GHC.Prim.coerce
          @(Bug.App Data.Functor.Compose.Compose f g a
            -> Bug.App Data.Functor.Compose.Compose f g a
               -> Bug.App Data.Functor.Compose.Compose f g a)
          @(Bug.Wat f g a -> Bug.Wat f g a -> Bug.Wat f g a)
          (Data.Semigroup.<>)
    Data.Semigroup.sconcat
      = GHC.Prim.coerce
          @(Data.List.NonEmpty.NonEmpty Bug.App Data.Functor.Compose.Compose f g a
            -> Bug.App Data.Functor.Compose.Compose f g a)
          @(Data.List.NonEmpty.NonEmpty Bug.Wat f g a -> Bug.Wat f g a)
          Data.Semigroup.sconcat
    Data.Semigroup.stimes
      = GHC.Prim.coerce
          @(forall (b :: TYPE GHC.Types.LiftedRep).
            GHC.Real.Integral b =>
            b
            -> Bug.App Data.Functor.Compose.Compose f g a
               -> Bug.App Data.Functor.Compose.Compose f g a)
          @(forall (b :: TYPE GHC.Types.LiftedRep).
            GHC.Real.Integral b => b -> Bug.Wat f g a -> Bug.Wat f g a)
          Data.Semigroup.stimes
```

I've reported this upstream as [#14578](https://ghc.haskell.org/trac/ghc/ticket/14578).

**Comment by RyanGlScott** [comment]
Fixed in 44433309011b33f03cb81eb1f163faa92a2a92fd.
